src/main/java/org/mockito/internal/PremainAttachAccess.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2024 Mockito contributors
+ * This program is made available under the terms of the MIT License.
+ */
+package org.mockito.internal;
+
+import java.lang.instrument.Instrumentation;
+import net.bytebuddy.ClassFileVersion;
+import net.bytebuddy.agent.ByteBuddyAgent;
+import net.bytebuddy.agent.Installer;
+
+import static org.mockito.internal.util.StringUtil.join;
+
+public class PremainAttachAccess {
+
+    public static Instrumentation getInstrumentation() {
+        // A Java agent is always added to the system class loader. If Mockito is executed from a
+        // different class loader we need to make sure to resolve the instrumentation instance
+        // from there, or fail the resolution, if this class does not exist on the system class
+        // loader.
+        ClassLoader systemClassLoader = ClassLoader.getSystemClassLoader();
+        Instrumentation instrumentation =
+                doGetInstrumentation(PremainAttach.class, "getInstrumentation", systemClassLoader);
+        if (instrumentation == null) {
+            try {
+                // At this point, we do not want to attempt dynamic attach but check the Byte Buddy
+                // agent as a fallback if the Mockito agent is not attached.
+                instrumentation =
+                        doGetInstrumentation(
+                                Installer.class, "getInstrumentation", systemClassLoader);
+            } catch (IllegalStateException ignored) {
+            }
+        }
+        if (instrumentation == null) {
+            if (ClassFileVersion.ofThisVm().isAtLeast(ClassFileVersion.JAVA_V21)) {
+                System.out.println(
```

Comment:
**thought:** I totally forgot after looking back at the code base maybe we can use `Plugins.getMockitoLogger().log(...)` here instead of the `System.out.println`.

Reply:
Current pr code uses st out, that said I like your idea on the default method.

Reply:
Reverted, see https://github.com/mockito/mockito/pull/3459#issuecomment-2396220498